Main.py

```diff
@@ -1,62 +1,249 @@
 """
+Cosette Ralowicz
+September 2024
+
 a gofundme donation caluclator to figure out what you should donate
 in order for the recipient to receive the full amount intended. Also to convert 
 between currency as needed
  """
+from abc import ABC, abstractmethod
+from datetime import datetime
+import sqlite3
+import requests
+import csv
+import json
+from datetime import datetime,timezone
+
+class AbstractCurrencyService(ABC):
+    """barebones of the api calling class and the backup data class"""
+    @abstractmethod
+    def get_currency_data(cur, con)->requests:
+        """ open api and update exchange rates in db and datetimes in csv """
+        pass
+    def updateDatetimes():
+       """update the csv with datetimes"""
+    pass
+    @abstractmethod
+    def compareDateTime():
+        """to see if it is time to call api again
+        compare current datetime (utc) with "time of next update" (utc) provided by the api which is saved in csv"""
+        pass
+    @abstractmethod
+    def getCurrency():
+        """get the user's input as to which currency the donation will be in"""
+        pass
+
+    @abstractmethod
+    def currencyRate(currency, cur):
+        """get the exchange rate for the user inputted currency from the db"""
+        pass
+    @abstractmethod
+    def check_for_currency(currency)->bool:
+        """check whether the user inputted currency exists within the db"""
+
+
+class APIService(AbstractCurrencyService):
+  
+    def get_currency_data(cur, con)->requests:
+        #get api response
+        url='https://open.er-api.com/v6/latest/USD'
+        response=requests.get(url)
+        #api success
+        if response.status_code==200:
+            data=response.json()
+  
+            rates=data['rates']
+            #update db 
+            try:
+                for country in rates.keys():
+                    update_query="""UPDATE rates SET rate = ? WHERE currency = ?"""
+                    #print(values)
+                    cur.execute(update_query, (rates[country],country ))
+                    con.commit()
+                    
+            except sqlite3.Error as error:
+                print("Failed to update MySQL table {}".format(error))
+            res=cur.execute("SELECT * FROM rates")
+
+            #update datetime csv 
+            nextUpdate=data['time_next_update_utc']
+            last_update=data['time_last_update_utc']
+            lastUpdate=datetime.strptime(last_update, '%a, %d %b %Y %X %z')
+            update=datetime.strptime(nextUpdate, '%a, %d %b %Y %X %z')
+            update_data={'time_last_update_utc':lastUpdate.strftime('%d %b %Y %X %z'),'time_next_update_utc': update.strftime('%d %b %Y %X %z') }
+            APIService.updateDatetimes(update_data)
+            return response
+        else:
+            return response
+        
+    def updateDatetimes(update_data):
+        with open('C:\\Users\\ckral\\oaf-psd-bootcamp\\update_datetime.csv', 'w') as currency_csv:
```

Comment:
In general, I would avoid using hardcoded file paths like this. Some helpful ways to avoid this:
1. You can choose to input/output files with a relative file path. https://stackoverflow.com/a/65174822
2. You can take in any file paths/directory as input from the user `--output_dir` and pass that along to your methods as needed.